lib/pmc_miller/reader.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require "pmc_miller/data"
+
+# PmcMiller
+module PmcMiller
+  ##
+  # Readers read in puppet-metrics-collector data
+  #
+  # Readers are associated with a specific puppet-metrics-collector directory.
+  # They can then assemble data for a specific key after they have been
+  # associated with the desired service.
+  class Reader
+    require "json"
+
+    attr_accessor :service
+
+    ##
+    # Creates a new reader that will use the given path as its data source.
+    #
+    # @param path [String] File system path to puppet-metrics-collector data
+    #
+    # @return [PmcMiller::Reader]
+    #
+    def initialize(path)
+      @path = path
+      @service = nil
+    end
+
+    ##
+    # Reads data associated with the given key for the assigned service of the
+    # reader.
+    #
+    # @param key [Symbol] Key to lookup data for
+    #
+    # @return [PmcMiller::Data] Data collection found
+    #
+    def read(key)
+      return unless %i[puppetserver puppetdb].include? @service
```

Comment:
Edge case: What would you expect the behavior to be for the following cases:
* The reader's service is assigned to a valid service that pmc failed to collect data for?  There would be no directory.  I would expect an empty dataset to be returned.
* The reader's service is assigned to a service that pmc does not support?  There would be no directory.  I would expect an exception to be thrown.

These cases lead me to create a whitelist of services rather than relying on the directory structure.  I am interested in your thoughts @RandellP .

Reply:
If PMC failed to collect data, there should still be a dir, it is created by the module in a way that is not connected to data collection I believe.  But I think it should raise rather than an empty dataset.  Cause there are multiple reason why it has no data, and the calling method can't figure out which it is from an empty dataset.  But a raised error can be caught, processed, and decisions made.  Like the key may not be in the data, there may be no data, there may be no dir...  heck, the jsons maybe invalid...  Lots of things could go wrong.  In most cases, the calling method will want to catch it and continue on in some way, but store up a list of issues to report out at the end.   All that said, I am not married to any part of what I said... :)